haystack/nodes/preprocessor/splitter.py

```diff
@@ -0,0 +1,738 @@
+                    unit_headlines.append(headline)
+                else:
+                    other_headlines.append(headline)
+
+            position_in_document += len(text)
+            headlines_to_assign = other_headlines
+
+            # Clone the meta from the parent document
+            unit_meta = deepcopy(document.meta)
+
+            # If the parent had headlines, but this unit happens not to have them, we assign an empty list
+            # If the parent never had a headlines field, we don't create it here.
+            if "headlines" in unit_meta and unit_meta["headlines"]:
+                unit_meta["headlines"] = unit_headlines
+
+            # Assing page number if required
+            if "page" in unit_meta:
+                del unit_meta["page"]
+            if add_page_number:
+                unit_meta["page"] = pages
+                pages += text.count("\f")
+
+            # Create the document
+            unit_document = Document(content=text, meta=unit_meta, id_hash_keys=document.id_hash_keys)
+            unit_documents.append(unit_document)
+
+        return unit_documents, units[1]
+
+    @staticmethod
+    def split_by_regex(pattern: str, text: str) -> Tuple[List[str], List[int]]:
+        """
+        Splits a long text into chunks based on a regex match.
+
+        :param pattern: The regex to split the text on.
+        :param text: The text to split.
+        :return: The list of splits, along with the length of the separators matched.
+        """
+        matches = [(match.start(), match.end()) for match in re.compile(pattern).finditer(text)]
+        if not matches:
+            return [text], [0]
+
+        if matches and not matches[-1][1] == len(text):
+            matches.append((len(text), len(text)))
+
+        units = []
+        offsets = []
+        for start_match, end_match in zip([(None, 0), *matches[:-1]], matches):
+            units.append(text[start_match[1] : end_match[1]])
+            offsets.append(end_match[1] - end_match[0])
+
+        return units, offsets
+
+    def split_by_sentence_tokenizer(self, text: str) -> Tuple[List[str], List[int]]:
+        """
+        Splits a given text with an NLTK sentence tokenizer, preserving all whitespace.
+
+        :param text: The text to tokenize.
+        :return: The tokenized text as a list of strings.
+        """
+        token_spans = self.sentence_tokenizer.span_tokenize(text)
+        return split_on_spans(text=text, spans=token_spans)
+
+    def split_by_nltk_word_tokenizer(self, text: str) -> Tuple[List[str], List[int]]:
+        """
+        Splits a given text with an NLTK word tokenizer, preserving all whitespace.
+
+        :param text: The text to tokenize.
+        :return: The tokenized text as a list of strings.
+        """
+        token_spans = self.tokenizer.span_tokenize(text)
+        return split_on_spans(text=text, spans=token_spans)
+
+    def split_by_dense_tokenizer(self, text: str) -> Tuple[List[str], List[int]]:
```

Comment:
Not sure about the naming here. I think something like `split_by_transformers_tokenizer` would be better.

Reply:
Good idea!